### ISSUE TITLE
A 6 x 6 linprog testcase: 3 methods x sparse / dense give different results

### ISSUE THREAD

**Issue by denis-bz** [scipy/scipy#10288]
#### A 6 x 6 linprog testcase from glpk: 3 methods x sparse / dense give different results

`transp-linprog.py` below is a 6 x 6 testcase of [scipy.optimize.linprog](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linprog.html)
from a GLPK example, transp. In outline:

    for method in ["interior-point", "revised simplex", "simplex"]:
        for sparse in [True, False]
            try:
                linprog( ... )

The 6 combinations give rather different results. grep `transp-linprog.log`:

    linprog  method: interior-point  sparse: True 
    The algorithm terminated successfully and determined that the problem is infeasible.
    final f: 0.649008  x: [0.711 0.57 0.515 0.707 0.557 0.549] 

    linprog  method: interior-point  sparse: False 
    The algorithm terminated successfully and determined that the problem is unbounded.
    final f: 0.649472  x: [0.712 0.57 0.515 0.708 0.557 0.55] 

    linprog  method: revised simplex  sparse: True 
    ValueError: shapes (1,12) and (1,12) not aligned: 12 (dim 1) != 1 (dim 0)

    linprog  method: revised simplex  sparse: False 
    final f: 153.675  x: [50 300 0 275 0 275] 

    linprog  method: simplex  sparse: True 
    ValueError: all the input arrays must have same number of dimensions

    linprog  method: simplex  sparse: False 
    final f: 153.675  x: [0 300 0 325 0 275] 

The full logfile and transp.* glpk files are under https://gist.github.com/denis-bz transp-linprog .

Transp is the smallest of the [GLPK](https://www.gnu.org/software/glpk) examples/, here made into a standalone testcase.

### Reproducing code example:

```
    #!/usr/bin/env python
    """ linprog testcase from transp.glp: all 3 methods, sparse / dense """

    from __future__ import division, print_function
    import sys
    import time
    import traceback  # cf. better-exceptions
    import numpy as np
    import scipy
    from scipy import sparse
    from scipy.optimize import linprog  # $scopt13/_linprog.py
        # https://docs.scipy.org/doc/scipy/reference/optimize.linprog-interior-point.html

    __version__ = "2019-06-10 june"
    __author_email__ = "denis-bz-py t-online.de"

    np.set_printoptions( threshold=10, edgeitems=5, linewidth=140,
            formatter = dict( float = lambda x: "%.3g" % x ))  # float arrays %.3g
    print( "\n" + 80 * "=" )
    print( "from", " ".join( sys.argv ), time.strftime( "%c" ))
    print( "versions: numpy %s  scipy %s   python %s " % (
            np.__version__, scipy.__version__ , sys.version.split()[0] ))

    #...............................................................................
    disp = True
    maxiter = 50
    tol = 1e-4  # default 1e-8

        # from glpk /examples/transp.glp glp_linprog --
    c = np.r_[ 0.225, 0.153, 0.162, 0.225, 0.162, 0.126 ]
    lb = np.r_[ 0, 0, 0, 0, 0, 0 ]
    ub = np.r_[ 1e10, 1e10, 1e10, 1e10, 1e10, 1e10 ]  # grr inf
    b = np.r_[ 1e10, 350, 600, -325, -300, -275 ]

    Adense = np.array(
        [[0.225, 0.153, 0.162, 0.225, 0.162, 0.126],
        [1, 1, 1, 0, 0, 0],
        [0, 0, 0, 1, 1, 1],
        [-1, 0, 0, -1, 0, 0],
        [0, -1, 0, 0, -1, 0],
        [0, 0, -1, 0, 0, -1]]
        )

        # to change these params, run this.py  a=1  b=None  'c = expr' ...
        # in sh or ipython --
    for arg in sys.argv[1:]:
        exec( arg )


    A = [Adense, sparse.csr_matrix( Adense )]
    bounds = np.c_[ lb, ub ]
    print( "c A: \n%s \n%s \nb: %s " % (c, Adense, b) )
    print( "params: disp %s  tol %g " % (disp, tol) )

    #...............................................................................
    for method in ["interior-point", "revised simplex", "simplex"]:
      for sparse in [True, False]:
        print( "\n" +  80 * "-" )
        print( "linprog  method: %s  sparse: %s " % (method, sparse) )
        options = dict( disp=disp, maxiter=maxiter, tol=tol,
                    sparse = sparse )  # rs, simplex OptimizeWarning
        try:
            res = linprog( A_ub=A[sparse], b_ub=b, c=c, bounds=bounds,
                    method=method, options=options )
            if not disp:
                print( "linprog:", res.message )
            print( "final f: %g  x: %s " % (res.fun, res.x) )
                # glpk: 50 300 0 275 0 275
        except (IndexError, ValueError):
            traceback.print_exc()
```

### Error messages:

above

### Scipy/Numpy/Python version information:

versions: numpy 1.16.4  scipy 1.3.0   python 3.7.3

**Comment by mdhaber** [comment]
Only interior point supports the sparse option, so that disposes of two cases.
Apparently there are multiple optimal vertices with the same value of the objective function, and both simplex methods find them, so there is no problem with those here.
You've set the tolerance to be very coarse. Does running the interior point methods with the default tolerance give the same result?

**Comment by mdhaber** [comment]
I didn't notice that the interior point methods returned a non-success status. The high upper bounds (1e10) might be causing scaling issues. Try replacing them with np.inf.

**Comment by denis-bz** [comment]
> Only interior point supports the sparse option

ok, but ValueError  ?

> does running the interior point methods with the default tolerance give the same result?

 py transp-linprog.py tol=1e-8:

linprog  method: interior-point  sparse: True 
    The algorithm terminated successfully and determined that the problem is infeasible.
    final f: 7.59467  x: [5.96 5.52 4.84 9.64 8.71 8.29] 

linprog  method: interior-point  sparse: False 
    The algorithm terminated successfully and determined that the problem is infeasible.
    final f: 7.611  x: [5.98 5.53 4.85 9.66 8.73 8.31] 

linprog  method: simplex  sparse: False 
    ValueError: Phase 1 of the simplex method failed to find a feasible solution. The pseudo-objective    function evaluates to 7.6e-06 which exceeds the required tolerance of 1e-08 for a solution to be considered 'close enough' to zero to be a basic solution. Consider increasing the tolerance to be greater than 7.6e-06. If this tolerance is unacceptably  large the problem may be infeasible.

> Try replacing them with np.inf

that would make sense, but
`_linprog_util.py _clean_inputs line 242`

       if not(np.isfinite(b_ub).all()):
            raise ValueError(
                "Invalid input for linprog: b_ub must not contain values "
                "inf, nan, or None")

I'd suggest first getting the doc aka spec right:  bounds arrays or array-like only, no tuples, no None s
=> simpler code,  `bounds_nx2`
=> compatible with some other scipy.optimizers
(except least_squares: 2-tuple of array_like)

For the doc on bounds, how about

<hr>
bounds: an n x 2 array of lower bounds and upper bounds 

To constrain variables `x[i]` between lower bounds and upper bound, `lb[i] <= x[i] <= ub[i]`,
make a `bounds` array with two columns `lb` and `ub`. For example,

   lb = np.zeros(n)            # n = number of variables, len(c)
   ub = np.inf * np.ones(n)    # inf: no upper bound
   lb[:10] = - np.inf
   ub[:10] = 1                 # constrain x[:10] <= 1, no lower bound
   bounds = np.c_[ lb, ub ]    # 2 columns, lb and ub
   lpres = linprog( ... bounds=bounds … )

Other forms of `bounds`:
`bounds=None` constrains all `x[i]` to be >= 0. This is the default.
`bounds` can be array-like, which means "can be converted to a numpy ndarray by `np.asarray`"

**Comment by denis-bz** [comment]
Scratch the above on bounds, silly: use the well-documented `class Bounds` in [_constraints.py](https://github.com/scipy/scipy/blob/master/scipy/optimize/_constraints.py)
like most other optimizers ?
Bounds in [_linprog_util.py](https://github.com/scipy/scipy/blob/master/scipy/optimize/_linprog_util.py) oughta be unified, not

	def _clean_inputs
		... grungy bounds code

	def _presolve
		...
		# Standard form for bounds (from _clean_inputs) is list of tuples
		# but numpy array is more convenient here
		# In retrospect, numpy array should have been the standard

**Comment by mdhaber** [comment]
Re: replacing `1e10` with `np.inf`, please replace the _bounds_ with `np.inf` and eliminate that first constraint entirely. I would agree that we should do this automatically (although it has never come up before). I just added automatic removal of constraints with infinite upper bound to #9671.

> ok, but ValueError ?
"Raised when an operation or function receives an argument that has the right type but an inappropriate value"
A suggestion for what it should be would be helpful.

Re: tolerance not fixing IP, I'm not surprised. I think it's the big numbers. Hopefully eliminating those big numbers will help.

Re: tolerance breaking `simplex`, this is not terribly surprising. After inclusion of `revised simplex`, `simplex` is only retained for backwards compatibility. @Kai-Striega might be interested to know about this, but there's probably nothing to be done about it as it's a tolerance issue.

Re: bounds specification, that is in #9671. I just added the possibility of specifying them as `Bounds` objects. When these were added to `minimize`, they weren't added to `linprog`.

**Comment by mdhaber** [comment]
> Bounds in _linprog_util.py oughta be unified
Yes, that's what the comment "In retrospect, numpy array should have been the standard" meant, that they should be converted to numpy arrays at the outset (in `_clean_inputs`) and `linprog` should have been structured to accept them like that. Unfortunately, that isn't the case - the rest of `linprog` does not take them as arrays - so at the end of `_presolve` they are converted _back_ to a list of tuples. 
```
# *sigh* - convert bounds back to their standard form (list of tuples)
# again, in retrospect, numpy array would be standard form
```
It's not pretty, but it is what it is, and changing it has been on the to do list for a long time.

PR welcome unless @Kai-Striega has already started?

**Comment by denis-bz** [comment]
Matt, folks.
  with inf bounds (checks in _linprog_util commented out) and tol=1e-8,
3 of the 5 smallest [GLPK](https://www.gnu.org/software/glpk) examples give

	/opt/local/mac/big/py3/lib/python3.7/site-packages/scipy/optimize/_linprog_ip.py:514: RuntimeWarning: invalid value encountered in double_scalars
	  rho_p = norm(r_p(x, tau)) / max(1, norm(r_p0))
	OptimizeWarning: Solving system with option 'cholesky':True failed. It is normal for this to happen occasionally, especially as the solution is approached. However, if you see this frequently, consider setting option 'cholesky' to False.
	OptimizeWarning: Solving system with option 'sym_pos':True failed. It is normal for this to happen occasionally, especially as the solution is approached. However, if you see this frequently, consider setting option 'sym_pos' to False.
	OptimizeWarning: Solving system with option 'sym_pos':False failed. This may happen occasionally, especially as the solution is approached. However, if you see this frequently, your problem may be numerically challenging. If you cannot improve the formulation, consider setting 'lstsq' to True. Consider also setting `presolve` to True, if it is not already.
	Numerical difficulties were encountered before the problem converged. Please check your problem formulation for errors, independence of linear equality constraints, and reasonable scaling and matrix condition numbers. If you continue to encounter this error, please submit a bug report.

GLPK solves them in ~ 0 seconds.

To run these I use a glp_sparse.py, test*py, shell scripts ...
which would certainly need some time with 1-1 contact to make pushbutton.
How could we proceed ?

Anyway in Germany it's time for a 🍺

**Comment by mdhaber** [comment]
I need to see the matrices to comment.
See #9636 for careful benchmarking against GLPK.

**Comment by mdhaber** [comment]
@denis-br can you post those matrices?

**Comment by Kai-Striega** [comment]
@denis-bz I refactored much of the simplex/interior-point code last year and while (to me) it's much better than it was, I'll be the first to admit that it's far from perfect. The linear programming module was written before ``_constraints`` were added and have never been updated since then. It is on the todo list (#9671) PRs or additional clean ups from there are welcome =).

@mdhaber I have started but don't think I'll finish anytime soon. I've had a friend show interest in it - but if someone else wants to start first they are free to go ahead with it.

**Comment by denis-bz** [comment]
Hi @Kai-Striega

should I use your newest _linprog*.py -- where can I find them ? (I'm not up to building all of scipy.)

Are we agreed that `inf` in `bounds` is ok ? I had to patch

    b[ ~ np.isfinite(b) ] = 1e20  #bz else NaNs in _indicators rho_A
    ...    
    rho_A = norm(c.T.dot(x) - b.T.dot(y)) / (tau + norm(b.T.dot(y)))  #bz inf in b -> NaN

Do umfpack / sksparse make a difference ? I have neither.
(Testing even 4 combinations looks like no fun, let alone versionitis.)

On more tests, I have a `glp_sparse.py`: .glp file -> scipy sparse
to provide a path from *any* GMPL (gnu Mathprog, a subset of AMPL) -> .glp -> scipy.
Running *.glp is easy in shell, less so in pure python.

cheers eh
  -- denis

**Comment by mdhaber** [comment]
> should I use your newest _linprog*.py -- where can I find them ? (I'm not up to building all of scipy.)

His changes have already been merged with the master branch. The version you find on the GitHub repository includes his changes.

I suggest that you look into getting your development environment set up if you intend to contribute code. There are ways around it, but it's not recommended.

> Are we agreed that `inf` in `bounds` is ok ?
 
`bounds` in its current state already supports `np.inf`. 
If you mean `b_ub`, the right hand side of the upper bound constraints, `np.inf` is _not_ ok. An upper bound on `b_ub` means there _is no constraint_, so the constraint should be removed entirely. The interior point algorithm will breaks down if you do not remove these trivial upper bound constraints.

> Do umfpack / sksparse make a difference ? I have neither.

They do (see [here](https://github.com/scipy/scipy/pull/10026) for performance comparison) but you don't necessarily have to test with all combinations locally. You would write unit tests and test them locally without SuiteSparse, and when you submit a PR the tests would be run with SuiteSparse.

> On more tests, I have a `glp_sparse.py`: .glp file -> scipy sparse
to provide a path from *any* GMPL (gnu Mathprog, a subset of AMPL) -> .glp -> scipy.
Running *.glp is easy in shell, less so in pure python.

I don't understand. Do you mean you have a system for converting GMPL to SciPy code? If so, please convert your tests to SciPy code and I can try to interpret them. It sounds like you are still including `np.inf` in the right hand side of upper bound constraints. Failure in that case is not a bug in the implementation; it is an inherent limitation of the algorithm combined with a lack of preprocessing to remove trivially-satisfied upper bound constraints. The solution is to add preprocessing to remove these unnecessary upper bound constraints.

**Comment by denis-bz** [comment]
RS works fine with `Inf` in `b_ub`, and I expected same input => same output from IP too --
principle of least astonishment.
Also `+- Inf` is standard in textbooks and in most? scipy optimizers to mean "no bound".
(To remove such rows, just

    jfinite = np.isfinite( b_ub )
    ninf = len(b_ub) - jfinite.sum()
    if ninf > 0:
        print( "linprog ip: removing %d rows of A_ub and b_ub where b_ub == inf" % ninf )
        A_ub = A_ub[jfinite]
        b_ub = b_ub[jfinite]

**Comment by mdhaber** [comment]
I agree that we should automatically remove the infinite right hand sides, so we need to leave this issue open until that's done. As you say, the fix is easy, so that would be a great first PR if you want to give it a shot.

Note that math textbooks also write A^-1 b, which in numerical computing should be interpretted as solve Ax = b (as opposed to calculate the inverse of A and multiply). I agree that we should make it easier for the user when we can, though.

**Comment by mdhaber** [comment]
@fletcheaston This would be another good issue to tackle because it would give an excuse to make a doc change and test building the docs locally.

The change to code should probably go in [`scipy/optimize/_linprog_util.py`](https://github.com/scipy/scipy/blob/master/scipy/optimize/_linprog_util.py), specifically in the `_clean_inputs` function around Line 242: before the check for all finite values, check for positive `np.inf` in `b_ub`. Remove all those entries and the corresponding rows in `A_ub`. 

(A_ub @ x <= b_ub are the inequality constraints of a linear programming problem. An entry of b_ub that is positive infinity corresponds with a constraint of the form a_1 x_1  + ... + a_n x_n <= inf, which is trivially satisfied. These constraints can and should be removed from the problem formulation as they do not change the solution of the problem, but they do make the problem larger and can cause numerical difficulties.)

In addition, let's modify the docstring in `scipy/optimize/_linprog.py`, `linprog` function, around line 208. Something like:
 ```
The inequality constraint vector. Each element represents an upper bound on the corresponding value of `A_ub @ x`. Elements `np.inf` and corresponding rows in `A_ub` will be automatically removed, as these constraints are trivially satisfied. For a faster, more reliable solution, manually remove constraints with exceptionally large upper bounds if they are known not to affect the solution.
```
Please test building the docs locally, then submit a PR and check the documentation built on the cloud. I left a mistake in the text above that won't render properly : )

**Comment by Kai-Striega** [comment]
I agree with Matt; we should keep the issue open for now, and that this
will be a good first issue for anyone that's interested in helping.

Should we still add this to the list of existing clean-ups?

**Comment by mdhaber** [comment]
@Kai-Striega it is already in the list of existing clean-ups.  
@fletcheaston is working on testing the revised contributor documentation (#10286) and I think he can use this as an issue.
@Kai-Striega Do you agree with the location of the code fix I suggested above? The doc change?

**Comment by Kai-Striega** [comment]
@mdhaber 

The code change: Positive infinity should be a valid upper bound, and so it should not raise an error. Saying that I think it would be much neater to keep cleaning/presolving as separate concerns in their respective functions. So eliminating constraints should be done in ``_presolve`` **not**  ``_clean_inputs``.

The docs: lgtm :+1:

**Comment by mdhaber** [comment]
OK @fletcheaston implement proposed code change around line 534, in `_presolve` instead of `_clean_inputs`.

That means that in `_clean_inputs`, you'll have to allow +/- inf to pass through without raising a `ValueError`. (I'm not sure, but `np.inf` and `float(inf)`/`math.inf` might be different, so be careful).

I'm guessing that `_presolve` doesn't have anything that will recognize that -inf as an upper bound makes the problem infeasible. If not, after dealing with +inf you'll have to check for -inf and if found return:
```
            status = 2
            message = ("The problem is (trivially) infeasible due to a"
                                "upper bound on a constraint being negative"
                                "infinity ")
            complete = True
            return (c, c0, A_ub, b_ub, A_eq, b_eq, bounds,
                    x, x0, undo, complete, status, message)
```

Other non-finite-number values should continue to raise the `ValueError`.

Sound good @Kai-Striega ?

**Comment by Kai-Striega** [comment]
@mdhaber must we let -inf pass? To me, an upper bound of negative infinity doesn't really make sense.  Also, at that point, we can expect `b` to be a numpy array of floats, so we won't need to worry about having math.inf.

**Comment by mdhaber** [comment]
@Kai-Striega We might as well let -inf pass. It's just as valid as +inf. It just makes the problem trivially infeasible.

**Comment by Kai-Striega** [comment]
@mdhaber and @fletcheaston I think there should be *two* changes.

- [ ] line 252 should allow -/+ inf pass, so only nans trigger an error. The error message should also be updated to reflect this.
- [ ] line 534 (approx)  should now remove the trivially infeasible constraints.

Further, I think it would be best to make the equality constraints match the upper bound constraints. That way the new behaviour stays consistent.

**Comment by mdhaber** [comment]
@Kai-Striega Such a problem could be either infeasible or unbounded, but I doubt the existing code will handle it properly. Do you really think it's worth the time to add the logic for determining which it is? 

This brings to mind that I was wrong before - negative infinity in the right hand side of the inequality constraint could be feasible (just unbounded). It's the same as negative infinity in the right hand side of an equality constraint.

I think we're going down a rabbit hole here. So I'm going to reverse my opinion and go with what you suggested originally - let +inf pass for RHS of inequality constraint, and raise `ValueError` for all other infinities/NaNs.

**Comment by Kai-Striega** [comment]
> Do you really think it's worth the time to add the logic for determining which it is?

TBH I'm not really sure. My intuition is that, we should either accept -/+ infinities as bounds - and treat them appropriately for all cases or exclude them entirely. OTOH:

> I think we're going down a rabbit hole here.

Agreed. In practice, I expect infinite bounds occur rarely - if at all.

> let +inf pass if it's in the RHS of an inequality constraint, but raise ValueError for all other infinities/NaNs.

This is probably the best compromise, the fix is straightforward, isn't very computationally expensive and covers the (I suspect) most common edge case.

> This LP solver does not solve that type of problem.

:+1: This could always be added in a later version, but for now, there are bigger priorities.
